File: backend/model/src/main/java/com/griddynamics/msd365fp/manualreview/model/dfp/Product.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

package com.griddynamics.msd365fp.manualreview.model.dfp;

import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.Data;

import java.io.Serializable;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategy.UpperCamelCaseStrategy.class)
public class Product implements Serializable {
    private String productId;
    private String productName;
    private String type;
    private String sku;
    private String category;
    private String market;
    private BigDecimal salesPrice;
    private BigDecimal salesPriceInUSD;
    private String currency;
    private BigDecimal currencyConversionFactor;
    @JsonProperty("COGS")
    private BigDecimal cogs;
    @JsonProperty("COGSInUSD")
    private BigDecimal cogsInUSD;
    private Boolean isRecurring;
    private Boolean isFree;
    private String language;
    private BigDecimal purchasePrice;
    private BigDecimal purchasePriceInUSD;
    private BigDecimal margin;
    private BigDecimal marginInUSD;
    private BigDecimal quantity;
    private Boolean isPreorder;
    private String shippingMethod;

    private Map<String, String> additionalParams = new HashMap<>();

    @JsonAnySetter
    public void setAdditionalParam(String name, String value) {
        additionalParams.put(name, value);
    }
    public void setAdditionalParams(Map<String, String> map) {
        additionalParams.putAll(map);
    }
}

File: frontend/src/models/item/purchase/product.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { ProductDTO } from '../../../data-services/api-services/models/product-dto';
import { BaseModel } from '../../misc';

export class Product extends BaseModel {
    productId: string = '';

    productName: string = '';

    type: string = '';

    sku: string = '';

    category: string = '';

    market: string = '';

    salesPrice: number = 0;

    salesPriceInUSD: number = 0;

    currency: string = '';

    COGS: number = 0;

    COGSInUSD: string = '';

    isRecurring: boolean = false;

    isFree: boolean = false;

    purchasePrice: number = 0;

    purchasePriceInUSD: number = 0;

    margin: number = 0;

    marginInUSD: number = 0;

    quantity: number = 0;

    isPreorder: boolean = false;

    shippingMethod: string = '';

    currencyConversionFactor: number = 0;

    language: string = '';

    fromDTO(product: ProductDTO) {
        const {
            ProductId,
            ProductName,
            Type,
            Sku,
            Category,
            Market,
            SalesPrice,
            Currency,
            COGS,
            IsRecurring,
            IsFree,
            Language,
            PurchasePrice,
            Margin,
            Quantity,
            IsPreorder,
            ShippingMethod,
            COGSInUSD,
            CurrencyConversionFactor,
            MarginInUSD,
            PurchasePriceInUSD,
            SalesPriceInUSD
        } = product;

        this.productId = this.define(this.productId, ProductId);
        this.productName = this.define(this.productName, ProductName);
        this.type = this.define(this.type, Type);
        this.sku = this.define(this.sku, Sku);
        this.category = this.define(this.category, Category);
        this.market = this.define(this.market, Market);
        this.salesPrice = this.define(this.salesPrice, SalesPrice);
        this.currency = this.define(this.currency, Currency);
        this.COGS = this.define(this.COGS, COGS);
        this.isRecurring = this.define(this.isRecurring, IsRecurring);
        this.isFree = this.define(this.isFree, IsFree);
        this.language = this.define(this.language, Language);
        this.purchasePrice = this.define(this.purchasePrice, PurchasePrice);
        this.margin = this.define(this.margin, Margin);
        this.quantity = this.define(this.quantity, Quantity);
        this.isPreorder = this.define(this.isPreorder, IsPreorder);
        this.shippingMethod = this.define(this.shippingMethod, ShippingMethod);
        this.COGSInUSD = this.define(this.COGSInUSD, COGSInUSD);
        this.currencyConversionFactor = this.define(this.currencyConversionFactor, CurrencyConversionFactor);
        this.marginInUSD = this.define(this.marginInUSD, MarginInUSD);
        this.purchasePriceInUSD = this.define(this.purchasePriceInUSD, PurchasePriceInUSD);
        this.salesPriceInUSD = this.define(this.salesPriceInUSD, SalesPriceInUSD);

        return this;
    }
}
